src/analyses/ai.h

```diff
@@ -354,7 +354,48 @@ class ai_baset
   virtual std::unique_ptr<statet> make_temporary_state(const statet &s)=0;
 };
 
-// domainT is expected to be derived from ai_domain_baseT
+/// Base class for static analyses. Any actual static analysis
+/// must (a) inherit from this class and (b) provide a domain class as a
+/// type argument, which must, in turn, inherit from \ref ai_domain_baset.
+///
+/// From a user's perspective, this class provides three main classes of
+/// functions:
+///
+/// 1. Running a static analysis, via
+///    \ref ai_baset#operator()(const irep_idt&,const goto_programt&, <!--
+///    --> const namespacet&),
+///    \ref ai_baset#operator()(const goto_functionst&,const namespacet&)
+///    and \ref ai_baset#operator()(const goto_modelt&)
+/// 2. Accessing the results of a static analysis, by looking up the result
+///    for a given location \p l using
+///    \ref ait#operator[](goto_programt::const_targett).
+/// 3. Outputting the results of the analysis; see
+///    \ref ai_baset#output(const namespacet&, const irep_idt&,
+///    const goto_programt&, std::ostream&)const et cetera.
+///
+/// A typical usage pattern would be to call the static analysis first,
+/// and use `operator[]` afterwards to retrieve the results. The fixed
+/// point algorithm used is a standard worklist algorithm; the current
+/// implementation is flow- and path-sensitive, but not context-sensitive.
+///
+/// From an analysis developer's perspective, an analysis is implemented by
+/// inheriting from this class (or, if a concurrency-sensitive analysis is
+/// required, from \ref concurrency_aware_ait), providing a class implementing
+/// the abstract domain as the type for the \p domainT parameter. Most of the
+/// actual analysis functions (in particular, the minimal element, the lattice
+/// join, and the state transformer) are supplied using \p domainT.
+///
+/// To control the analysis in more detail, you can also override the following
+/// methods:
+/// - \ref ait#initialize(const irep_idt&, const goto_programt&),
+///   \ref ait#initialize(const irep_idt&,
+///   const goto_functionst::goto_functiont&) and
+///   \ref ait#initialize(const goto_functionst&), for pre-analysis
+///   initialization
+/// - \ref ait#finalize(), for post-analysis cleanup.
+///
+// For the <!-- --> trick above, see
+// https://stackoverflow.com/questions/46744573/
```

Comment:
Maybe that should instead go in the commit message? (It's very nice indeed, we've got a few way-too-long-lines.)